### PR TITLE
planner/cascades: add ImplementationRule for TopN

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

This PR adds ImplementationRule for TopN.

### What is changed and how it works?

1. LogicalTopN can be implemented as PhysicalTopN or PhysicalLimit with required property.

2. Fix a bug about the cost of enfored sort.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

### CODE REVIEW

File: planner/cascades/implementation_rules.go [hunk 1]
```diff
@@ -272,3 +277,60 @@ func (r *ImplLimit) OnImplement(expr *memo.GroupExpr, reqProp *property.Physical
...
+// ImplTopNAsLimit is the implementation rule which implements LogicalTopN
+// as PhysicalLimit with required order property.
+type ImplTopNAsLimit struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTopNAsLimit) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	topN := expr.ExprNode.(*plannercore.LogicalTopN)
+	_, canUseLimit := plannercore.GetPropByOrderByItems(topN.ByItems)
+	return plannercore.MatchItems(prop, topN.ByItems) && canUseLimit
```
Comment (lamxTyler) [inline planner/cascades/implementation_rules.go]: `canUseLimit && plannercore.MatchItems(prop, topN.ByItems)` to take the benefits of short circuit evaluation?

File: planner/cascades/implementation_rules.go [hunk 2]
```diff
@@ -272,3 +277,60 @@ func (r *ImplLimit) OnImplement(expr *memo.GroupExpr, reqProp *property.Physical
...
+// ImplTopNAsLimit is the implementation rule which implements LogicalTopN
+// as PhysicalLimit with required order property.
+type ImplTopNAsLimit struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTopNAsLimit) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	topN := expr.ExprNode.(*plannercore.LogicalTopN)
+	_, canUseLimit := plannercore.GetPropByOrderByItems(topN.ByItems)
+	return canUseLimit && plannercore.MatchItems(prop, topN.ByItems)
```
Comment (lzmhhh123) [inline planner/cascades/implementation_rules.go]: It seems no need to check `MatchItems` again?

Comment (francis0407) [reply planner/cascades/implementation_rules.go]: `canUseLimit` checks if all of the order by items are columns. `MatchItems` checks if this TopN can satisfy the required property.

**Pull request merged**